src/train.py

```diff
@@ -16,67 +16,248 @@
+            print("Lengths", len_mask_idx.shape)
+            per += compute_per(character.to(DEVICE), text_pred, text_len.to(DEVICE), len_mask_idx)
+            n_iters += 1
+
+    model.train()
+
+    return per/n_iters, losses
+
+
+def optimizer_step(loss, model, optimizer, args):
+
+    # Take a optimizer step!
+    optimizer.zero_grad()
+    loss.backward()
+    if args.grad_clip > 0.0:
+        nn.utils.clip_grad_norm_(model.parameters(), args.grad_clip)
+    optimizer.step()
+    return loss.detach().cpu().item()
+
+
+def autoencoder_step(model, batch):
+    """
+    Compute and return the loss for autoencoders
+    """
+    character, mel, mel_input, pos_text, pos_mel, text_len = batch
+    character, mel, mel_input = character.to(DEVICE), mel.to(DEVICE), mel_input.to(DEVICE)
+    mel_gold, char_gold = mel.detach(), character.detach().permute(1, 0)
+
+    text_pred = model.text_ae(character).permute(1, 2, 0)
+    pred, stop_pred = model.speech_ae(mel, mel_input)
+    
+    # TODO: add end length loss
+    s_ae_loss = speech_loss(mel_gold, pos_mel, pred, stop_pred)
+    t_ae_loss = text_loss(gold_char, pred_char)
+    return t_ae_loss, s_ae_loss
+
+
+def supervised_step(model, batch):
+    character, mel, mel_input, pos_text, pos_mel, text_len = batch
+    character, mel, mel_input = character.to(DEVICE), mel.to(DEVICE), mel_input.to(DEVICE)
+    mel_gold, char_gold = mel.detach(), character.detach().permute(1, 0)
+
+    pred, stop_pred = model.tts(character, mel_input)
+    text_pred = model.asr(character, mel).permute(1, 2, 0)
+    
+    tts_loss = speech_loss(mel_gold, pos_mel, pred, stop_pred)
+    asr_loss = text_loss(gold_char, pred_char)
+    return asr_loss, tts_loss
+
+
+def crossmodel_step(model, batch):
+    #NOTE: not sure if this will fail bc multiple grads on the model...
+    character, mel, mel_input, pos_text, pos_mel, text_len = unsupervised_batch
+    character, mel, mel_input = character.to(DEVICE), mel.to(DEVICE), mel_input.to(DEVICE)
+    mel_gold, char_gold = mel.detach(), character.detach().permute(1, 0)
+
+    # Do speech!
+    pred, stop_pred = model.cm_speech_in(mel, mel_input)
+    s_cm_loss = speech_loss(mel_gold, pos_mel, pred, stop_pred)
+
+    # Now do text!
+    text_pred = model.cm_text_in(character).permute(1, 2, 0)
+    t_cm_loss = text_loss(gold_char, pred_char)
+    return t_cm_loss, s_cm_loss
+
+def train(args):
+    set_seed(args.seed)
+
+    print("#### Getting Dataset ####")
+    supervised_train_dataset = get_dataset('labeled_train.csv')
+    unsupervised_train_dataset = get_dataset('unlabeled_train.csv')
+    val_dataset = get_dataset('val.csv')
+    full_train_dataset = get_dataset('full_train.csv')
+
+    # init models and optimizers
+    text_m = TextRNN(args).to(DEVICE)  
+    speech_m = SpeechRNN(args).to(DEVICE)
+    discriminator = None
+    model = UNAST(text_m, speech_m, discriminator)
+    optimizer = torch.optim.Adam(model.parameters(), lr=args.lr,  weight_decay=1e-5)
+    best = 100
```

Comment:
Would it be possible to create a separate function for initializing the model (and no need to add lots of `args` test to it yet):

```python3
def initialize_model(args, device):
    # We can add schedulers to this later if we need to

    # initialize model
    text_m, speech_m, discriminator = None, None, None
    if args.model_type == 'rnn':
        text_m = TextRNN(args).to(device)
        speech_m = SpeechRNN(args).to(device)
    if args.model_type == 'transformer':
        ...
    if args.use_discriminator:
        discriminator = ...
    model = UNAST(text_m, speech_m, discriminator)

    # initialize optimizer
    optimizer = None
    if args.optim_type == 'adam':
        optimizer = torch.optim.Adam(model.parameters(), lr=args.lr, weight_decay=1e-5)

    return model, optimizer
```

So that we can reuse the `train()` function with other models:

```python3
def train(args, model, optimizer):
    ...
```

And call `initialize_model(args, DEVICE)` in `"__main__"`.